File: TinyScript/java/src/main/java/paser/ast/FunctionArgs.java
package paser.ast;

import lexer.Token;
import paser.util.ParseException;
import paser.util.PeekTokenIterator;

/**
 * @author MFine
 * @version 1.0
 * @date 2021/10/27 23:34
 **/
public class FunctionArgs extends ASTNode {
    public FunctionArgs(ASTNode parent) {
        super(parent);
        this.label = "args";
    }

    public static ASTNode parse(ASTNode parent, PeekTokenIterator it) throws ParseException {
        FunctionArgs functionArgs = new FunctionArgs(parent);

        while (it.peek().isType()) {
            Token type = it.next();
            Variable var = (Variable) Factor.parse(parent, it);
            assert var != null;
            var.setTypeLexeme(type);
            functionArgs.addChild(var);
            if (!it.peek().getValue().equals(")")) {
                it.nextMatch(",");
            }
        }

        return functionArgs;
    }
}

File: TinyScript/js/src/main/js/parser/ast/FunctionArgs.ts
import ASTNode from "./ASTNode";
import ASTNodeTypes from "./ASTNodeTypes";
import PeekTokenIterator from "../PeekTokenIterator";
import Variable from "./Variable";

export  default class FunctionArgs extends ASTNode {

    constructor() {
        super(ASTNodeTypes.FUCTION_ARGS, 'args');
    }

    static parse(it: PeekTokenIterator): FunctionArgs {
        const {Factor} = require('./index');
        const functionArgs = new FunctionArgs();
        while (it.peek()?.isType()) {
            const type = it.next();
            const variable = Factor.parse(it) as Variable;
            functionArgs.addChild(variable);
            variable.typeLexeme = type;
            if (it.peek()?.value !== ')') {
                it.nextMatch(',');
            }
        }
        return functionArgs;
    }
}
